File: sdk/python/veritas_sdk/rules.py
"""Rule predicate checks, hand-mirrored from core/src/circuits/*.rs.

Only the banking-basel-iii transaction-threshold rule is implemented here
today, matching core/src/circuits/banking_basel_iii.rs -- see that file's
module docstring for why this is the one rule out of the compliance
mapping docs that's actually been implemented (it's the one each doc
flags as tractable near-term).
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

RULE_ID_BANKING_BASEL_III = "banking-basel-iii"


class RuleViolation(Exception):
    """Raised when a rule's predicate does not hold for the given input.
    Mirrors VeritasError::RuleViolation in core/src/errors.rs.
    """


@dataclass(frozen=True)
class TransactionThresholdInput:
    transaction_amount_minor: int
    risk_adjusted_threshold_minor: int
    customer_id_hash: bytes  # must be exactly 32 bytes

    def __post_init__(self) -> None:
        if len(self.customer_id_hash) != 32:
            raise ValueError(
                f"customer_id_hash must be exactly 32 bytes, got {len(self.customer_id_hash)}"
            )


def check_transaction_threshold(input_: TransactionThresholdInput) -> None:
    """Raises RuleViolation if the transaction exceeds the customer's
    risk-adjusted threshold. Mirrors TransactionThresholdRule::check in
    core/src/circuits/banking_basel_iii.rs exactly (including the
    less-than-or-equal boundary: exactly-at-threshold passes).
    """
    if input_.transaction_amount_minor > input_.risk_adjusted_threshold_minor:
        raise RuleViolation(
            f"{RULE_ID_BANKING_BASEL_III}: transaction_amount_minor "
            f"{input_.transaction_amount_minor} exceeds risk_adjusted_threshold_minor "
            f"{input_.risk_adjusted_threshold_minor}"
        )


def canonical_bytes(input_: TransactionThresholdInput) -> bytes:
    """Must byte-for-byte match
    TransactionThresholdRule::canonical_bytes in Rust: two little-endian
    u64s followed by the raw 32-byte hash. This is exactly the encoding
    core/tests/vectors/banking-basel-iii.json is meant to pin down cross-
    language -- see tests/test_vectors.py in this package.
    """
    return (
        struct.pack("<Q", input_.transaction_amount_minor)
        + struct.pack("<Q", input_.risk_adjusted_threshold_minor)
        + input_.customer_id_hash
    )

File: sdk/typescript/src/rules.ts
// Rule predicate checks, hand-mirrored from core/src/circuits/*.rs.
// Only banking-basel-iii's transaction-threshold rule is implemented here
// today -- see core/src/circuits/banking_basel_iii.rs's module docstring
// for why that's the one rule actually implemented across all languages.
//
// u64 fields use `bigint` (not `number`) throughout: JS numbers lose
// precision above 2^53, which real transaction-amount-in-minor-units
// values can exceed. Silently downgrading to `number` here would be a
// real correctness bug, not just a style choice.

export const RULE_ID_BANKING_BASEL_III = "banking-basel-iii";

export class RuleViolation extends Error {
  constructor(message: string) {
    super(message);
    this.name = "RuleViolation";
  }
}

export interface TransactionThresholdInput {
  transactionAmountMinor: bigint;
  riskAdjustedThresholdMinor: bigint;
  /** Must be exactly 32 bytes. */
  customerIdHash: Uint8Array;
}

function assertValidInput(input: TransactionThresholdInput): void {
  if (input.customerIdHash.length !== 32) {
    throw new RangeError(
      `customerIdHash must be exactly 32 bytes, got ${input.customerIdHash.length}`,
    );
  }
  if (input.transactionAmountMinor < 0n || input.riskAdjustedThresholdMinor < 0n) {
    throw new RangeError("amounts must be non-negative");
  }
}

/**
 * Throws RuleViolation if the transaction exceeds the customer's
 * risk-adjusted threshold. Mirrors TransactionThresholdRule::check in
 * core/src/circuits/banking_basel_iii.rs exactly, including the boundary:
 * exactly-at-threshold passes.
 */
export function checkTransactionThreshold(input: TransactionThresholdInput): void {
  assertValidInput(input);
  if (input.transactionAmountMinor > input.riskAdjustedThresholdMinor) {
    throw new RuleViolation(
      `${RULE_ID_BANKING_BASEL_III}: transaction_amount_minor ` +
        `${input.transactionAmountMinor} exceeds risk_adjusted_threshold_minor ` +
        `${input.riskAdjustedThresholdMinor}`,
    );
  }
}

/**
 * Must byte-for-byte match TransactionThresholdRule::canonical_bytes in
 * Rust: two little-endian u64s followed by the raw 32-byte hash. Pinned
 * cross-language by core/tests/vectors/banking-basel-iii.json (see
 * tests/vectors.test.ts in this package).
 */
export function canonicalBytes(input: TransactionThresholdInput): Uint8Array {
  assertValidInput(input);
  const buf = new Uint8Array(8 + 8 + 32);
  const view = new DataView(buf.buffer);
  view.setBigUint64(0, input.transactionAmountMinor, /* littleEndian */ true);
  view.setBigUint64(8, input.riskAdjustedThresholdMinor, true);
  buf.set(input.customerIdHash, 16);
  return buf;
}
